gunicorn/_reloader.py

```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -
+#
+# This file is part of gunicorn released under the MIT license.
+# See the NOTICE for more information.
+
+import os.path
+import re
+import sys
+
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+from watchdog.observers.polling import PollingObserver
+
+
+class Reloader(object):
+
+    def __init__(self, extra_files=None, interval=1, callback=None,
+                 engine='auto'):
+        super(Reloader, self).__init__()
+        if engine == 'auto' or engine == 'inotify':
+            self._observer = Observer()
+        else:
+            assert engine == 'poll'
+            self._observer = PollingObserver(timeout=interval)
+        self._observer.daemon = True
+        self._handler = ReloadingEventHandler(callback)
+        self._dirs = set()
+
+        if extra_files is not None:
+            for filename in extra_files:
+                self.add_extra_file(filename)
+
+    def add_extra_file(self, filename):
+        dirname = os.path.dirname(filename)
+        self._observer.schedule(self._handler, dirname)
+
+    def get_dirs(self):
```

Comment:
I don't like gunicorn's approach attempt at being smart about reloading, I think it would be better to take directories to watch recursively (and add directories which are created), as the current watching leaves a lot of [holes](https://github.com/benoitc/gunicorn/issues/1562#issuecomment-326415960). Then the user can have their project directory watched for changes (including resources like templates and static content).